### PR TITLE
update kafka license

### PR DESCRIPTION
**Pull request by pjfanning**
* follow up to https://github.com/apache/incubator-pekko-connectors-kafka/pull/89
* put Kafka license header in code

### CODE REVIEW

**Review by He-Pin — APPROVED** [review]
lgtm

File: cluster-sharding/src/main/scala/org/apache/pekko/kafka/cluster/sharding/KafkaClusterSharding.scala [hunk 1]
```diff
@@ -12,6 +12,25 @@
  * Copyright (C) 2016 - 2020 Lightbend Inc. <https://www.lightbend.com>
  */
 
+// license header for
```
Comment (mdedetrich) [inline cluster-sharding/src/main/scala/org/apache/pekko/kafka/cluster/sharding/KafkaClusterSharding.scala:15]: Don't you mean from?

**Review by mdedetrich — COMMENTED** [review]
I found one thing but this seems a bit odd to me, making a reference that we take a header from somewhere feels wrong.

It seems like instead we should be just copying the header (I'm this ASF standard header) and just mention code is derived from Apache Kafka?

**Comment by pjfanning** [comment]
> I found one thing but this seems a bit odd to me, making a reference that we take a header from somewhere feels wrong.
> 
> It seems like instead we should be just copying the header (I'm this ASF standard header) and just mention code is derived from Apache Kafka?

The code from Akka has the headers that it has, we can't remove them. The code here is a mix of Akka code and Kafka code. 

We need all the headers. LEGAL-626, Lightbend and Kafka.

File: cluster-sharding/src/main/scala/org/apache/pekko/kafka/cluster/sharding/KafkaClusterSharding.scala (hunk 1 above)
Comment (pjfanning) [reply cluster-sharding/src/main/scala/org/apache/pekko/kafka/cluster/sharding/KafkaClusterSharding.scala]: changed

**Comment by mdedetrich** [comment]
> We're getting -1s in votes. Provide too much information, no -1. Provide too little, we get -1. I'm erring on the side of maximising the chances of getting releases approved.

I understand the context and I am not saying we shouldn't do anything and/or provide less info, just that the current info we are providing language wise seems wrong. Even with the current `license header from` text it sounds like we are taking the header from other source which legally speaking is a no no (you never "take" headers from another source, the headers are coupled with source code  and you automatically bring it over).

**Comment by pjfanning** [comment]
> > We're getting -1s in votes. Provide too much information, no -1. Provide too little, we get -1. I'm erring on the side of maximising the chances of getting releases approved.
> 
> I understand the context and I am not saying we shouldn't do anything and/or provide less info (the header obviously needs to be there). Its just that some of the current info we are providing language wise seems wrong, i.e. with the current `license header from` text it sounds like we are taking the header from other source which legally speaking is a no no (you never "take" headers from another source, the headers are coupled with source code and you automatically bring it over).
> 
> EDIT: To put this another away, I have never seen comments pertaining to "header coming from/taken from x" and if I actually see that text rather than being legally helpful it has the other effect of raining red flags.

For me, it is equally confusing to have so many license headers without providing some context as to why we have so many - including 2 separate Apache headers.

**Comment by pjfanning** [comment]
`based on Apache Kafka's [[org.apache.kafka.clients.producer.internals.DefaultPartitioner]]` appears in multiple places

Moving the comment about the license header into scaladoc where the `based on` comment above is not a good idea (imo).

Adding it as a non-scaladoc comment along side those comments means a lot of duplication - having one comment near the license header just seems the tidiest solution (to me).

I have not seen any rules that forbid us from adding informational comments around the license headers to help users understand why they are there.

**Comment by mdedetrich** [comment]
> I have not seen any rules that forbid us from adding informational comments around the license headers to help users understand why they are there.

Thats not the issue, my concern is that when someones with no context looks at this file, even with the clarifying comment, it looks incredibly suspicious because in essence we have duplicated headers. Our bespoke ASF header is meant to grant the same effect as the standard ASF header + including a note that it was derived from Akka **and** we have another ASF header (because its from Apache Kafka) because....?

Isn't a much better solution just modifying our bespoke ASF header to

```scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * license agreements; and to You under the Apache License, version 2.0:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * This file is part of the Apache Pekko project, which was derived from Akka
 * and Apache Kafka.
 */
 ```
 
 (note the `and Apache Kafka`) part. Its clear, provides just as much info and most critically doesn't look suspect.

**Comment by pjfanning** [comment]
> > I have not seen any rules that forbid us from adding informational comments around the license headers to help users understand why they are there.
> 
> Thats not the issue, my concern is that when someone with no context looks at this file, even with the clarifying comment, it looks incredibly suspicious because in essence we have duplicated headers. Our bespoke ASF header is meant to grant the same effect as the standard ASF header + including a note that it was derived from Akka **and** yet we have another standard ASF header (due to it being based on code from Apache Kafka) because....?
> 
> Isn't a much better solution just modifying our bespoke ASF header to
> 
> ```scala
> /*
>  * Licensed to the Apache Software Foundation (ASF) under one or more
>  * license agreements; and to You under the Apache License, version 2.0:
>  *
>  *   https://www.apache.org/licenses/LICENSE-2.0
>  *
>  * This file is part of the Apache Pekko project, which was derived from Akka
>  * and Apache Kafka https://github.com/apache/kafka/blob/451fff8937c626a187371b8fcdf50cfb5177d38f/clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java
>  */
> ```
> 
> (note the `and Apache Kafka` part at the end). Its clear, provides just as much info and most critically doesn't look suspect. You can even specify the full filename if you want, i.e.
> 
> ```scala
> /*
> * Licensed to the Apache Software Foundation (ASF) under one or more
> * license agreements; and to You under the Apache License, version 2.0:
> *
> *   https://www.apache.org/licenses/LICENSE-2.0
> *
> * This file is part of the Apache Pekko project, which was derived from Akka
> * and Apache Kafka (specifically 
> * https://github.com/apache/kafka/blob/451fff8937c626a187371b8fcdf50cfb5177d38f/clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java)
> */
> ```

Could you raise an issue in the LEGAL Jira project for this? https://issues.apache.org/jira/

**Comment by mdedetrich** [comment]
Will do

**Comment by mdedetrich** [comment]
Ticket created at https://issues.apache.org/jira/browse/LEGAL-648

**Comment by pjfanning** [comment]
closing based on https://issues.apache.org/jira/browse/LEGAL-648

**Pull request closed without merging**